exec credential provider: wire in cluster info (superset of #91192)

ankeesler opened the pull request:
* This PR should contain a strict superset of what @enj did in: https://github.com/kubernetes/kubernetes/pull/91192.
  * I opened this PR since I don't think I can push to @enj's branch in https://github.com/kubernetes/kubernetes/pull/91192.
  * I created my new branch from the tip of @enj's branch in https://github.com/kubernetes/kubernetes/pull/91192, so it should have his same commits, except now I am marked as the committer since I rebased `master` on top of his stuff.
* The KEP for this exec credential feature set is here: https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/541-external-credential-providers/README.md
* This PR should contain functionality agreed upon in the KEP updates
  * KEP update 1: https://github.com/kubernetes/enhancements/pull/1927
  * KEP update 2: https://github.com/kubernetes/enhancements/pull/2023
* The original k/e proposal for this feature set is here: https://github.com/kubernetes/enhancements/issues/541

Signed-off-by: Andrew Keesler <akeesler@vmware.com>

/kind feature
/sig auth
/milestone v1.20

```release-note
client-go credential plugins can now be passed in the current cluster information via the KUBERNETES_EXEC_INFO environment variable.
```

```docs
Need to update https://kubernetes.io/docs/reference/access-authn-authz/authentication/#client-go-credential-plugins
```

enj committed:
exec credential provider: wire in cluster info

Signed-off-by: Monis Khan <mok@vmware.com>

enj committed:
Generated conversions

Signed-off-by: Monis Khan <mok@vmware.com>

enj committed:
Generated bazel

Signed-off-by: Monis Khan <mok@vmware.com>

ankeesler commented:
TODO:
- [ ] fix `LoadExecCredential` tests
- [ ] update `exec.Authenticator` tests with optional `Cluster` change
- [ ] fix indentation of struct field in `clientcmd/api/types.go` and `clientcmd/api/v1/types.go`
- [ ] add references from PR to KEP design

fedebongio commented:
/cc @cheftako @DangerOnTheRanger
/remove-sig api-machinery

ankeesler commented on staging/src/k8s.io/client-go/tools/auth/exec/exec_test.go:
```diff
@@ -0,0 +1,189 @@
+}
+
+// TODO: are there global test helpers that do this?
+func match(t *testing.T, expected, actual interface{}) {
```
Are there well-known kube testing utilities that do this? The best I could do was find that `diff` package...

liggitt replied:
cmp.Diff(a, b) is what we normally use

ankeesler replied:
https://github.com/kubernetes/kubernetes/pull/95489/commits/bd4af0df52b7da2a838795ccf9a8760a0316da48

ankeesler commented on staging/src/k8s.io/client-go/tools/auth/exec/exec_test.go:
```diff
@@ -0,0 +1,189 @@
+	utilruntime.Must(corev1.AddToScheme(scheme))
+}
+
+func TestLoadExecCredential(t *testing.T) {
```
I manually tested `LoadExecCredentialFromEnv` using a one-off exec plugin and a `kubectl` built from source. Should this manual test be converted to an automatic test and checked into the repo somewhere?

liggitt replied:
I think the individual unit tests are probably sufficient

ankeesler commented:
This PR is ready for review. Helpful context can be found in the first comment. https://github.com/kubernetes/kubernetes/pull/95489#issue-501604799

/assign @liggitt

ankeesler commented on staging/src/k8s.io/client-go/tools/auth/exec/exec.go:
```diff
@@ -0,0 +1,130 @@
+package exec
```
How does a kubernetes developer find out about helper packages such as this one? Is there some sort of documentation that needs to be updated (besides the website)?

ankeesler replied:
Partially answered by https://github.com/kubernetes/kubernetes/pull/95489/files#r510141353. Resolving.

fedebongio commented:
/remove-sig api-machinery

dims commented:
/ok-to-test

ankeesler commented:
thanks @dims !

dims commented:
/priority important-soon
/triage accepted
/milestone v1.20

dims commented:
/assign @smarterclayton

liggitt commented on staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1alpha1/conversion.go:
```diff
@@ -0,0 +1,26 @@
+)
+
+func Convert_clientauthentication_ExecCredentialSpec_To_v1alpha1_ExecCredentialSpec(in *clientauthentication.ExecCredentialSpec, out *ExecCredentialSpec, s conversion.Scope) error {
+	return autoConvert_clientauthentication_ExecCredentialSpec_To_v1alpha1_ExecCredentialSpec(in, out, s)
```
add a comment indicating this intentionally omits the Cluster field which is only supported in newer version

ankeesler replied:
https://github.com/kubernetes/kubernetes/pull/95489/commits/bd4af0df52b7da2a838795ccf9a8760a0316da48

liggitt commented on staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1beta1/conversion.go:
```diff
@@ -17,10 +17,10 @@ limitations under the License.
+	"k8s.io/client-go/pkg/apis/clientauthentication"
 )
 
 func Convert_clientauthentication_ExecCredentialSpec_To_v1beta1_ExecCredentialSpec(in *clientauthentication.ExecCredentialSpec, out *ExecCredentialSpec, s conversion.Scope) error {
```
add a comment indicating this intentionally excludes response and interactive, which were only supported in v1alpha1

ankeesler replied:
https://github.com/kubernetes/kubernetes/pull/95489/commits/bd4af0df52b7da2a838795ccf9a8760a0316da48

liggitt commented on staging/src/k8s.io/client-go/rest/config.go:
```diff
@@ -87,7 +87,7 @@ type Config struct {
 
 	// Exec-based authentication provider.
-	ExecProvider *clientcmdapi.ExecConfig
+	Exec Exec
```
can we avoid adding this nesting level? this is an extremely widely used struct and changes here are are disruptive

ankeesler replied:
I remember you and @enj talking about this in https://github.com/kubernetes/kubernetes/pull/91192/files#r444297938.

Of the two options you presented in that comment, I think I like `ClusterConfig runtime.Object to ExecConfig` the best, to try to avoid adding another field to `rest.Config`?

ankeesler replied:
https://github.com/kubernetes/kubernetes/pull/95489/commits/bd4af0df52b7da2a838795ccf9a8760a0316da48

liggitt commented on staging/src/k8s.io/client-go/rest/config.go:
```diff
@@ -603,7 +642,10 @@ func CopyConfig(config *Config) *Config {
-		ExecProvider:        config.ExecProvider,
+		Exec: Exec{
+			ExecProvider: config.Exec.ExecProvider,
+			Config:       config.Exec.Config,
```
should this deep copy the config.Exec.Config if it is non-nil?

ankeesler replied:
I think that sounds reasonable...

ankeesler replied:
https://github.com/kubernetes/kubernetes/pull/95489/commits/bd4af0df52b7da2a838795ccf9a8760a0316da48

ankeesler replied:
BTW, I started to add a test to verify that we deep copy this member, but then backed off since it seemed ugly and I didn't know if we needed it. I can add it back if you think that would be healthier. It was something to the tune of this.
```go
		if actual.ExecProvider != nil {
			if expected.ExecProvider == nil {
				t.Fatal("CopyConfig dropped ExecProvider field")
			}
			if actual.ExecProvider.Config != nil {
				if expected.ExecProvider.Config == nil {
					t.Fatal("CopyConfig dropped ExecProvider.Config field")
				}
				if actual.ExecProvider.Config `is the object same as` expected.ExecProvider.Config {
					t.Fatalf("CopyConfig did not deep copy ExecProvider.Config field (%p == %p)", actual.ExecProvider.Config, expected.ExecProvider.Config)
				}
			}
		}

```

liggitt commented on staging/src/k8s.io/client-go/rest/transport.go:
```diff
@@ -89,12 +91,22 @@ func (c *Config) TransportConfig() (*transport.Config, error) {
-	if c.ExecProvider != nil {
-		provider, err := exec.GetAuthenticator(c.ExecProvider)
+	if c.Exec.ExecProvider != nil {
+		caData, err := dataFromSliceOrFile(c.CAData, c.CAFile)
```
only load this and fail if the provider requested cluster data, right?

ankeesler replied:
https://github.com/kubernetes/kubernetes/pull/95489/commits/bd4af0df52b7da2a838795ccf9a8760a0316da48

liggitt commented on staging/src/k8s.io/client-go/rest/transport.go:
```diff
@@ -89,12 +91,22 @@ func (c *Config) TransportConfig() (*transport.Config, error) {
+		if err != nil {
+			return nil, fmt.Errorf("failed to load CA bundle for execProvider: %v", err)
+		}
+		cluster := clientauthentication.Cluster{
```
isn't this missing other bits like proxyURL and insecure? is there a unit test that should have caught this?

ankeesler replied:
https://github.com/kubernetes/kubernetes/pull/95489/commits/bd4af0df52b7da2a838795ccf9a8760a0316da48

whoops. yeah, you're right - we _should_ be passing through those other fields.

i am scratching my head for where we could put this unit test. i don't see a unit test in the `rest` package for this code. i also don't see any test higher up that verifies that the initialization of an exec plugin is done correctly.

i could add a unit test in `rest`, but it seems non-trivial to get the cluster info that we pass to `exec.GetAuthenticator`. i was tempted to add a test to `k8s.io/client-go/plugin/pkg/auth/exec`, but those unit tests are pretty self-contained to that package's functionality (which totally makes sense).

best i could come up with is to add a helper function that maps from `rest.Config` to `clientauthentication.Cluster`.

the missing test coverage that you called out still exists, because if i completely remove the call to the new helper function, then all the `rest` unit tests still pass. do you have any better ideas?

liggitt commented on staging/src/k8s.io/client-go/tools/clientcmd/api/v1/types.go:
```diff
@@ -214,6 +214,12 @@ type ExecConfig struct {
+	// ProvideClusterInfo determines whether or not to provide cluster information,
+	// which could potentially contain very large CA data, to this exec plugin as a
+	// part of the KUBERNETES_EXEC_INFO environment variable. By default, it is set
+	// to false.
```
can point to the helper package with methods for dealing with the envvar here

ankeesler replied:
https://github.com/kubernetes/kubernetes/pull/95489/commits/bd4af0df52b7da2a838795ccf9a8760a0316da48

liggitt commented on staging/src/k8s.io/client-go/tools/auth/exec/exec.go:
```diff
@@ -0,0 +1,147 @@
+	return obj, restConfig, nil
+}
+
+// makeKubeconfig makes a kubeconfig out of the provided cluster.
+func makeKubeconfig(cluster *clientauthentication.Cluster) *clientcmdapi.Config {
```
this bit is a little awkward, but I'm not sure we have a better way to unify the two structs... is there a fuzzing test that ensures any new fields added to Cluster will have to be added here and round-tripped correctly?

ankeesler replied:
This PR adds `"k8s.io/client-go/pkg/apis/clientauthentication/v1beta1".TestCluster` which ensures that `clientauthenticationv1beta1.Cluster` has all the same fields and tags as `clientcmdv1.Cluster`, but we don't have any fuzzing test to ensure we are copying over the fields we want here in `makeKubeconfig`. So I can add a fuzzing test for that in this (`k8s.io/client-go/tools/auth/exec`) package.

ankeesler replied:
https://github.com/kubernetes/kubernetes/pull/95489/commits/bd4af0df52b7da2a838795ccf9a8760a0316da48

See my comment above about unit testing strategy...

ankeesler commented on staging/src/k8s.io/client-go/tools/clientcmd/api/v1/conversion.go:
```diff
@@ -172,3 +172,9 @@ func Convert_Map_string_To_runtime_Object_To_Slice_v1_NamedExtension(in *map[str
 }
+
+func Convert_api_ExecConfig_To_v1_ExecConfig(in *api.ExecConfig, out *ExecConfig, s conversion.Scope) error {
+	// This intentionally omits the in.Config field which is an internal-only field for storing
```
I was getting the following error when doing codegen, so I added this manual conversion.
```
E1026 16:32:58.380934   33931 conversion.go:750] Warning: could not find nor generate a final Conversion function for k8s.io/client-go/tools/clientcmd/api.ExecConfig -> k8s.io/client-go/tools/clientcmd/api/v1.ExecConfig
E1026 16:32:58.381097   33931 conversion.go:751]   the following fields need manual conversion:
E1026 16:32:58.381105   33931 conversion.go:753]       - Config
```

ankeesler commented:
/test pull-kubernetes-node-e2e

liggitt commented on staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1beta1/types_test.go:
```diff
@@ -0,0 +1,97 @@
+// We want Cluster to offer the same knobs as clientcmdv1.Cluster to allow someone to connect to
+// the kubernetes API. This test should fail if a new field is added to one of the structs without
+// updating the other.
+func TestCluster(t *testing.T) {
```
This test looks good for catching gaps in both directions.

would this make more sense in staging/src/k8s.io/client-go/tools/auth/exec which already has visibility to both the `k8s.io/client-go/tools/clientcmd/api` and `k8s.io/client-go/pkg/apis/clientauthentication` API packages? it seems like it would pair well with the test that ensures we can construct rest.Config objects with fidelity from the envvar.

ankeesler replied:
Sure, sounds reasonable to me. I was nervous to move it to `.../tools/...` because I didn't know if there was a convention around testing API structs as close to their declaration as possible, but I totally agree with you that `k8s.io/client-go/tools/clientcmd/api` seems like the natural place for this since the dependency graph works out nicely.

ankeesler replied:
502dd3fcf0d

liggitt commented on staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1beta1/types_test.go:
```diff
@@ -0,0 +1,97 @@
+				t.Parallel()
+				clientcmdField, ok := clientcmdType.FieldByName(execField.Name)
+				if !ok {
+					t.Errorf("unknown field '%s'", execField.Name)
+				} else if execField.Type != clientcmdField.Type {
+					t.Error("type mismatch:", execField.Type, "!=", clientcmdField.Type)
+				} else if execField.Tag != clientcmdField.Tag {
+					t.Error("tag mismatch:", execField.Tag, "!=", clientcmdField.Tag)
```
be explicit about what the expected action is here... typically, it will be to fix the exec cluster config type

ankeesler replied:
77da453042f

liggitt commented on staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1beta1/types_test.go:
```diff
@@ -0,0 +1,97 @@
+				t.Parallel()
+				execField, ok := execType.FieldByName(clientcmdField.Name)
+				if !ok {
+					t.Errorf("unknown field '%s'", clientcmdField.Name)
```
be explicit about what the expected action is here... typically, it will be to add the new field to the exec cluster config type

ankeesler replied:
77da453042f

liggitt commented on staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1beta1/types_test.go:
```diff
@@ -0,0 +1,97 @@
+				if !ok {
+					t.Errorf("unknown field '%s'", clientcmdField.Name)
+				} else if clientcmdField.Type != execField.Type {
+					t.Error("field mismatch:", clientcmdField.Type, "!=", execField.Type)
+				} else if clientcmdField.Tag != execField.Tag {
+					t.Error("tag mismatch:", clientcmdField.Tag, "!=", execField.Tag)
```
be explicit about what the expected action is here... typically, it will be to fix the exec cluster config type

ankeesler replied:
77da453042f

liggitt commented on staging/src/k8s.io/client-go/rest/exec.go:
```diff
@@ -0,0 +1,58 @@
+// false, then this function will return an empty clientauthenticationapi.Cluster.
+func ConfigToExecCluster(config *Config) (*clientauthenticationapi.Cluster, error) {
+	if config.ExecProvider == nil || !config.ExecProvider.ProvideClusterInfo {
+		return &clientauthentication.Cluster{}, nil
```
this is inconsistent with the API doc that indicates the cluster field will be nil if ProvideClusterInfo is false. It looks like we only call this in one non-test location; should we put the burden of doing this check on the caller instead and assume if we are called that `config.ExecProvider != nil && config.ExecProvider.ProvideClusterInfo == true`?

ankeesler replied:
09323e5c557

This will likely change once we figure out where to put this thing...

liggitt commented on staging/src/k8s.io/client-go/rest/transport.go:
```diff
@@ -94,7 +94,11 @@ func (c *Config) TransportConfig() (*transport.Config, error) {
 	}
 
 	if c.ExecProvider != nil {
-		provider, err := exec.GetAuthenticator(c.ExecProvider)
```
would it make more sense to change this to `exec.GetAuthenticator(c)` and move ConfigToExecCluster to the `k8s.io/client-go/plugin/pkg/client/auth/exec` package which already has visibility to `k8s.io/client-go/tools/clientcmd/api` and `k8s.io/client-go/pkg/apis/clientauthentication`?

ankeesler replied:
hopefully this is resolved by https://github.com/kubernetes/kubernetes/pull/95489#discussion_r513892294

liggitt commented on staging/src/k8s.io/client-go/tools/auth/exec/exec.go:
```diff
@@ -0,0 +1,150 @@
+	return obj, restConfig, nil
+}
+
+func execClusterToConfig(cluster *clientauthentication.Cluster) (*rest.Config, error) {
```
going all the way through a mock kubeconfig struct seems pretty roundabout. why are we not translating these fields directly? we have the following methods which should round-trip (modulo loading CA file and resolving the proxy URL):

* execClusterToConfig: exec cluster -> rest.Config
* ConfigToExecCluster: rest.Config -> exec cluster

I expected both of these to live in similar places and have a test verifying they could round-trip with each other

ankeesler replied:
> going all the way through a mock kubeconfig struct seems pretty roundabout. why are we not translating these fields directly?

Yup, totally agree, it seems silly. I started to write a manual conversion from `exec cluster` to `rest.Config`, but then I realized I was rewriting a lot of kubeconfig loading logic. To avoid code duplication, and possibly missing out on future `rest.Config`-creation features/bug fixes, I elected to create the mock kubeconfig. This is certainly a loosely held opinion, though.

> I expected both of these to live in similar places

Agreed. This was another "where should these code elements go" moment. If we put `ConfigToExecCluster`, per your suggestion above[0], in `k8s.io/client-go/plugin/pkg/client/auth/exec`, then I guess it would follow existing dependency graph edges, now that I think of it. I thought that `plugin` packages were only supposed to be imported into `main` packages, and the like, for registration. However, I'm just now realizing that `k8s.io/client-go/rest` depends on `k8s.io/client-go/plugin/pkg/client/auth/exec`, and `k8s.io/client-go/tools/auth/exec` depends on `k8s.io/client-go/rest`, so therefore `k8s.io/client-go/tools/auth/exec` already has a dependency on `k8s.io/client-go/plugin/pkg/client/auth/exec`. As long as you are OK with me importing `k8s.io/client-go/plugin/pkg/client/auth/exec` into `k8s.io/client-go/tools/auth/exec`, I'll do that.

[0] https://github.com/kubernetes/kubernetes/pull/95489#discussion_r513538814

liggitt replied:
> I'm just now realizing that k8s.io/client-go/rest depends on k8s.io/client-go/plugin/pkg/client/auth/exec, and k8s.io/client-go/tools/auth/exec depends on k8s.io/client-go/rest, so therefore k8s.io/client-go/tools/auth/exec already has a dependency on k8s.io/client-go/plugin/pkg/client/auth/exec.

That was my thinking as well

liggitt replied:
> I realized I was rewriting a lot of kubeconfig loading logic

Really? I expected a pretty straight copy of the fields… there's no user or context info. It seems like it would be less work and more straightforward than the field by field copy you are doing here

ankeesler replied:
I'll go with the straight copy of the fields. Could certainly have been that I was making up stories in my head...

ankeesler replied:
Oh nooooo...I think if we followed through with this plan, `plugin/pkg/client/auth/exec` would depend on `rest` (because of return type below), which already depends on `plugin/pkg/client/auth/exec`...

`func ExecClusterToConfig(cluster *clientauthentication.Cluster) (*rest.Config, error)`

ankeesler replied:
OK new bad idea: put both `ExecClusterToConfig` and `ConfigToExecCluster` in `rest`, roundtrip test them there, continue to inject the `clientauthentcation.Cluster` into `exec.GetAuthenticator`, and call `ExecClusterToConfig` from `tools/auth/exec`. I think that is the only way the dependency graph will line up.

liggitt replied:
I think you're right. That's not that bad... put them in their own file and there's a certain symmetry to them

ankeesler replied:
gave this a shot here: 930bfd161a2

liggitt commented on staging/src/k8s.io/client-go/rest/exec.go:
```diff
@@ -0,0 +1,58 @@
+
+	var proxyURL string
+	if config.Proxy != nil {
+		url, err := config.Proxy(nil)
```
shouldn't this be passing in the config host? also, Proxy can return a nil url with no error

```suggestion
hostURL, err := url.Parse(config.Host)
...
proxyURL, err := config.Proxy(hostURL)
...
if proxyURL != nil {
  ... = proxyURL.String()
}
```

ankeesler replied:
The `proxyURL` nil check makes sense, but I believe `config.Proxy()` accepts a `*http.Request`.

```golang
	Proxy func(*http.Request) (*url.URL, error)
```

Maybe `config.Proxy(&http.Request{URL: hostURL})`? This is kinda a tricky conversion, because we are essentially converting from a dynamic proxy URL (`rest.Config.Proxy`) to a static proxy URL (`clientauthentication.Client`). I would feel bad passing a half-baked `http.Request` to `config.Proxy`.

liggitt replied:
mm, true. probably `http.NewRequest("", config.Host, nil)` to get the same normalization we'd get for actual requests.

ankeesler replied:
0750ed9c614

ankeesler commented on staging/src/k8s.io/client-go/tools/clientcmd/api/types.go:
```diff
@@ -215,6 +215,35 @@ type ExecConfig struct {
+	// such as the audience.  This field allows the per cluster config to be directly
+	// specified with the cluster info.  Using this field to store secret data is not
+	// recommended as one of the prime benefits of exec plugins is that no secrets need
+	// to be stored directly in the kubeconfig.
```
bd62001e034

ankeesler commented on staging/src/k8s.io/client-go/rest/config_test.go:
```diff
@@ -525,8 +535,9 @@ func TestConfigStringer(t *testing.T) {
-					Env:  []clientcmdapi.ExecEnvVar{{Name: "secret", Value: "s3cr3t"}},
+					Args:   []string{"secret"},
+					Env:    []clientcmdapi.ExecEnvVar{{Name: "secret", Value: "s3cr3t"}},
+					Config: &runtime.Unknown{Raw: []byte("super secret password")},
```
38db9c4b6b3

ankeesler commented on staging/src/k8s.io/client-go/tools/clientcmd/api/types.go:
```diff
@@ -237,7 +266,11 @@ func (c ExecConfig) String() string {
+	if c.Config != nil {
+		config = "runtime.Object(--- REDACTED ---)"
+	}
+	return fmt.Sprintf("api.AuthProviderConfig{Command: %q, Args: %#v, Env: %s, APIVersion: %q, ProvideClusterInfo: %t, Config: %s}", c.Command, args, env, c.APIVersion, c.ProvideClusterInfo, config)
```
b46f46b3002

ankeesler commented on staging/src/k8s.io/client-go/tools/auth/exec/types_test.go:
```diff
@@ -0,0 +1,125 @@
+// allow someone to connect to the kubernetes API. This test should fail if a new field is added to
+// one of the structs without updating the other.
+//
+// TODO: when clientauthenticationv1 is born, we should probably have a similar test.
```
I'm all ears for suggestions on how we can make sure we don't forget to do this when we got to GA. I know we can add a TODO item somewhere in GitHub, but I would love to be able to add something in code.

liggitt replied:
you can check if a ExecCredential is registered for v1 into the scheme created in staging/src/k8s.io/client-go/tools/auth/exec/exec.go and fail with a "TODO: add field fuzzing test" if so (can test it with a check for v1beta1 to make sure the check works)

ankeesler replied:
excellent. 56d9b8fda37

ankeesler commented on staging/src/k8s.io/client-go/tools/auth/exec/exec_test.go:
```diff
@@ -0,0 +1,205 @@
+/*
```
How would we feel about a test that validates that the `rest.Config` returned from `LoadExecCredential` can actually be used to make a REST call? I worry that we will change something in `rest.ExecClusterToConfig` and forget to validate that `LoadExecCredential` still returns a REST config that is valid for HTTP calls. Perhaps this is being too wary...

liggitt commented on staging/src/k8s.io/client-go/rest/exec.go:
```diff
@@ -0,0 +1,85 @@
+		}
+		url, err := config.Proxy(req)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get proxy URL: %w", err)
```
```suggestion
			return nil, fmt.Errorf("failed to get proxy URL for execProvider: %w", err)
```

ankeesler replied:
b4ce993f7ce

liggitt commented on staging/src/k8s.io/client-go/rest/exec.go:
```diff
@@ -0,0 +1,85 @@
+	if config.Proxy != nil {
+		req, err := http.NewRequest("", config.Host, nil)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create empty http request: %w", err)
```
```suggestion
			return nil, fmt.Errorf("failed to create proxy URL request for execProvider: %w", err)
```

ankeesler replied:
b4ce993f7ce

liggitt commented on staging/src/k8s.io/client-go/rest/transport.go:
```diff
@@ -94,7 +95,15 @@ func (c *Config) TransportConfig() (*transport.Config, error) {
 
 	if c.ExecProvider != nil {
-		provider, err := exec.GetAuthenticator(c.ExecProvider)
+		var cluster clientauthentication.Cluster
```
make this a pointer, don't dereference, and pass a pointer to GetAuthenticator?

```suggestion
		var cluster *clientauthentication.Cluster
```

ankeesler replied:
I agree this would clean up the code here in `transport.go`. I was trying to avoid passing a pointer to `GetAuthenticator` that could be nil or non-nil, so that `GetAuthenticator` wouldn't need to worry about limping around with a nil pointer. On the other hand, passing this struct by pointer would alleviate some copying of a `Cluster` up the stack. I will go ahead and update the `GetAuthenticator` code, and friends, to take a pointer. It doesn't look like we touch that `cluster` struct very much right now.

ankeesler replied:
bfe3c622358

liggitt commented on staging/src/k8s.io/client-go/tools/auth/exec/exec.go:
```diff
@@ -0,0 +1,110 @@
+
+	restConfig, err := rest.ExecClusterToConfig(execCredential.Spec.Cluster)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot create config: %w", err)
```
```suggestion
		return nil, nil, fmt.Errorf("cannot create rest.Config: %w", err)
```

ankeesler replied:
962c4f64655

liggitt commented on staging/src/k8s.io/client-go/tools/auth/exec/exec.go:
```diff
@@ -0,0 +1,110 @@
+	}
+
+	if execCredential.Spec.Cluster == nil {
+		return nil, nil, errors.New("exec credential does not contain cluster information")
```
```suggestion
		return nil, nil, errors.New("ExecCredential does not contain cluster information")
```

ankeesler replied:
962c4f64655

liggitt commented on staging/src/k8s.io/client-go/tools/auth/exec/exec_test.go:
```diff
@@ -0,0 +1,205 @@
+}
+
+func init() {
+	utilruntime.Must(corev1.AddToScheme(scheme))
```
why is this needed? modifying the non-test package `scheme` in a test-only file doesn't seem like a great idea... I really want to exercise the methods that use the scheme the way non-test code will

ankeesler replied:
I used a `Pod` down below to validate unmarshalling a `runtime.Object` that wasn't a `clientauthentication.ExecCredential`. I'll use a different `runtime.Object` for this so we don't need to modify the `scheme` in the test.

ankeesler replied:
962c4f64655

liggitt commented on staging/src/k8s.io/client-go/tools/auth/exec/types_test.go:
```diff
@@ -0,0 +1,137 @@
+// TestV1beta1ClusterTypesAreSynced after we promote ExecCredential to GA. It should start failing
+// when someone adds support for the v1 type to this package.
+func TestV1ClusterTypesAreSynced(t *testing.T) {
+	if _, ok := scheme.AllKnownTypes()[schema.GroupVersionKind{
```
this is fine, we could also iterate through AllKnownTypes and fail if we find any `clientauthenticationv1beta1.SchemeGroupVersion.Group / ExecCredential` group kinds outside the versions we already expect (v1alpha1 and v1beta1)

ankeesler replied:
Makes sense to me. I guess that would protect against stuff like `v1beta2` sneaking in.

ankeesler replied:
962c4f64655

ankeesler commented on staging/src/k8s.io/client-go/rest/exec.go:
```diff
@@ -0,0 +1,85 @@
+// clientauthenticationapi.Cluster. The returned Config will be anonymous (i.e., it will not have
+// any authentication-related fields set).
+func ExecClusterToConfig(cluster *clientauthentication.Cluster) (*Config, error) {
+	var proxy func(*http.Request) (*url.URL, error)
```
In our kubeconfig loading logic, we ensure that the proxy URL' scheme is valid (http/https/socks5)[0]. Furthermore, we have some logic around the `rest.Config.Host`[1]. I didn't include any of that in here, hoping that the API contract for this function could be something like "I'll do my best to copy over the fields one-for-one, but you are on your own if any of those fields are invalid".

[0] https://github.com/kubernetes/kubernetes/blob/2729b8e375143434fc4977fe49eaea572567dac3/staging/src/k8s.io/client-go/tools/clientcmd/helpers.go#L38
[1] https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/clientcmd/client_config.go#L171

liggitt replied:
yeah, that's fine

liggitt commented:
looks like `hack/update-bazel.sh` is needed as well

liggitt commented:
this lgtm, go ahead and squash review commits into layers that make sense

ankeesler committed:
exec credential provider: ProvideClusterInfo and kubeconfig shadow

- The main idea here is that we want to 1) prevent potentially large CA
  bundles from being set in an exec plugin's environment and 2) ensure
  that the exec plugin is getting everything it needs in order to talk to
  a cluster.
- Avoid breaking existing manual declarations of rest.Config instances by
  moving exec Cluster to kubeconfig internal type.
- Use client.authentication.k8s.io/exec to qualify exec cluster extension.
- Deep copy the exec Cluster.Config when we copy a rest.Config.

Signed-off-by: Andrew Keesler <akeesler@vmware.com>

ankeesler committed:
exec credential provider: k8s.io/client-go/tools/auth/exec helper

Exec plugin implementations should be able to call
LoadExecCredentialFromEnv() in order to get everything they need to
operate (i.e., cluster information (as long as it is passed in) and
optionally per-cluster configuration).

Signed-off-by: Andrew Keesler <akeesler@vmware.com>

ankeesler commented:
squashed into 2 major commits. i kept @enj's commits intact since i wanted to leave his as author on those. i tried to pull apart c4299d15d5289768808034676858e76a177eeae5 more, but changing struct fields were caused some widespread changes.

ankeesler commented:
/retest

ankeesler commented:
Note: i've been using this exec plugin to run end-to-end tests manually: https://gist.github.com/ankeesler/2502d73d164164fc16b1a54bf4607423.

liggitt reviewed (commented):
/approve

sorry, noticed one doc issue on the types, lgtm after that is fixed

ankeesler committed:
exec credential provider: exec -> client.authentication.k8s.io/exec

Hopefully we've fixed all of these references now...

Signed-off-by: Andrew Keesler <akeesler@vmware.com>

ankeesler commented on staging/src/k8s.io/client-go/pkg/apis/clientauthentication/types.go:
```diff
@@ -106,7 +122,7 @@ type Cluster struct {
 	//     ...
 	//     extensions:
-	//     - name: exec  # reserved extension name for per cluster exec config
+	//     - name: client.authentication.k8s.io/exec  # reserved extension name for per cluster exec config
```
good catch, 409f22832a9

ankeesler commented on staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1beta1/types.go:
```diff
@@ -87,7 +103,7 @@ type Cluster struct {
 	//     ...
 	//     extensions:
-	//     - name: exec  # reserved extension name for per cluster exec config
+	//     - name: client.authentication.k8s.io/exec  # reserved extension name for per cluster exec config
```
good catch, 409f22832a9

liggitt commented:
/lgtm
/approve

ankeesler commented:
/test pull-kubernetes-e2e-kind-ipv6

ankeesler commented:
/test pull-kubernetes-e2e-gce-100-performance

ankeesler commented:
/test pull-kubernetes-node-e2e

ankeesler commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

dims commented:
test ... ignore me!
